internal/timeline_check_runner.go

```diff
@@ -0,0 +1,98 @@
+package internal
+
+import (
+	"github.com/wal-g/storages/storage"
+	"github.com/wal-g/tracelog"
+	"github.com/wal-g/wal-g/utility"
+)
+
+// TimelineCheckRunner is used to verify that the current timeline
+// is the highest among the storage timelines
+type TimelineCheckRunner struct {
+	currentTimeline  uint32
+	storageFileNames []string
+}
+
+func (check TimelineCheckRunner) Name() string {
+	return "TimelineCheck"
+}
+
+func NewTimelineCheckRunner(rootFolder storage.Folder, currentSegment WalSegmentDescription) (TimelineCheckRunner, error) {
+	walFolder := rootFolder.GetSubFolder(utility.WalPath)
+	storageFileNames, err := getFolderFilenames(walFolder)
+	if err != nil {
+		return TimelineCheckRunner{}, err
+	}
+	return TimelineCheckRunner{currentTimeline: currentSegment.Timeline, storageFileNames: storageFileNames}, nil
+}
+
+func (check TimelineCheckRunner) Run() (WalVerifyCheckResult, error) {
+	highestTimeline := tryFindHighestTimelineId(check.storageFileNames)
+	return newTimelineCheckResult(check.currentTimeline, highestTimeline), nil
+}
+
+func (check TimelineCheckRunner) Type() WalVerifyCheckType {
+	return WalVerifyTimelineCheck
+}
+
+type TimelineCheckResult struct {
+	CurrentTimelineId        uint32 `json:"current_timeline_id"`
+	HighestStorageTimelineId uint32 `json:"highest_storage_timeline_id"`
+}
+
+// newTimelineCheckResult check produces the WalVerifyCheckResult with status:
+// StatusOk if current timeline id matches the highest timeline id found in storage
+// StatusWarning if could not determine if current timeline matches the highest in storage
+// StatusFailure if current timeline is not equal to the highest timeline id found in storage
+func newTimelineCheckResult(currentTimeline, highestTimeline uint32) WalVerifyCheckResult {
+	result := WalVerifyCheckResult{
+		Status: StatusWarning,
+		Details: TimelineCheckResult{
+			CurrentTimelineId:        currentTimeline,
+			HighestStorageTimelineId: highestTimeline,
+		},
+	}
+	if highestTimeline > 0 {
```

Comment:
We get a warning when highestTimeline == 0. Is it supposed?

Reply:
yes, highest timeline == 0 means that something is not right. For example, there are no WAL segments in storage